### PR TITLE
[FEATURE][WIP] Support for 3D visualisation of TIN as terrain (issue #32240)

### PR DESCRIPTION
**Pull request by vcloarec**
Support TIN as terrain with map texture. 
For discussion.
There are still some work to do, especially fix texture issue in the tile boundary.
tin_3d

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [x] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

### CODE REVIEW

**Comment by nyalldawson** [comment]
Nice work!!

File: src/3d/terrain/qgsterraintileloader_p.h [hunk 1]
```diff
@@ -59,6 +60,12 @@ class QgsTerrainTileLoader : public QgsChunkLoader
     //! Gives access to the terain entity
     QgsTerrainEntity *terrain() { return mTerrain; }
 
+    void setExtentMapCrs( const QgsRectangle &extent )
+    {
+      qDebug() << extent.toRectF();
```
Comment (nyalldawson) [inline src/3d/terrain/qgsterraintileloader_p.h]: Don't forget to remove this prior to merge

File: src/3d/terrain/qgstindemterraingenerator.cpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+void QgsTinDemTerrainGenerator::updateTilingScheme()
+{
+  if ( mLayer )
+    mTerrainTilingScheme = QgsTilingScheme( mLayer->extent(), mLayer->crs() );
+  else
+    mTerrainTilingScheme = QgsTilingScheme();
+}
+
+QgsTinDemTerrainTileLoader::QgsTinDemTerrainTileLoader( QgsTerrainEntity *terrain, QgsChunkNode *node, QgsMeshLayer *layer ):
+  QgsTerrainTileLoader( terrain, node ), mLayer( layer ),
```
Comment (nyalldawson) [inline src/3d/terrain/qgstindemterraingenerator.cpp]: QGIS code convention is for initialization lists to follow this format:

```
MyClass::MyClass( int a, int b )
    : mA( a )
    , mB( b )
{

}

```

File: src/3d/terrain/qgstindemterraingenerator.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+class QgsTinDemTerrainTileLoader: public QgsTerrainTileLoader
+{
+  public:
+    QgsTinDemTerrainTileLoader( QgsTerrainEntity *terrain, QgsChunkNode *node, QgsMeshLayer *layer );
+
+    Qt3DCore::QEntity *createEntity( Qt3DCore::QEntity *parent ) override;
+
+  private:
+    QgsMeshLayer *mLayer;
```
Comment (nyalldawson) [inline src/3d/terrain/qgstindemterraingenerator.h]:
```suggestion
    QgsMeshLayer *mLayer = nullptr;
```

File: src/3d/terrain/qgstindemterraingenerator.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    QgsTriangularMeshTile mMeshTile;
+};
+
+
+class _3D_EXPORT QgsTinDemTerrainGenerator: public QgsTerrainGenerator
+{
+  public:
+    QgsTinDemTerrainGenerator() {}
```
Comment (nyalldawson) [inline src/3d/terrain/qgstindemterraingenerator.h]:
```suggestion
    QgsTinDemTerrainGenerator() = default;
```

File: src/3d/terrain/qgstindemterraingenerator.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    ~ QgsTinDemTerrainGenerator() override {}
```
Comment (nyalldawson) [inline src/3d/terrain/qgstindemterraingenerator.h]: Just remove this, no point in an empty destructor here

File: src/3d/terrain/qgstindemterraintilegeometry_p.cpp [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+    QVector3D normal = mesh.vertexNormalVector( i );
+    normal = QVector3D( normal.x(), -normal.y(), normal.z() * vertScale );
+    normal.normalized();
+
+    *fptr++ = normal.x();
+    *fptr++ = normal.z();
+    *fptr++ = normal.y();
+  }
+
+  qDebug() << "return buffer bytes";
```
Comment (nyalldawson) [inline src/3d/terrain/qgstindemterraintilegeometry_p.cpp]: Remove

File: src/3d/terrain/qgstindemterraintilegeometry_p.cpp [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+
+  for ( int i = 0; i < faces; ++i )
+  {
+
+    const QgsMeshFace &face = mesh.triangle( i );
+    for ( int i = 0; i < 3; ++i )
+      *indexPtr++ = quint32( face.at( i ) );
+  }
+
+  qDebug() << "return index bytes";
```
Comment (nyalldawson) [inline src/3d/terrain/qgstindemterraintilegeometry_p.cpp]: Remove (others too)

File: src/3d/terrain/qgstindemterraintilegeometry_p.cpp [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+{
+  mZmax = std::numeric_limits<float>::min();
+  mZmin = std::numeric_limits<float>::max();
+  mFaces = QVector<int>::fromList( mTriangularMesh->faceIndexesForRectangle( mExtent ) );
+  mVertices.clear();
+  mLocalIndexFromGlobalIndex.clear();
+
+  int localIndex = 0;
+  mRealExtent.setMinimal();
+  for ( auto f : mFaces )
```
Comment (nyalldawson) [inline src/3d/terrain/qgstindemterraintilegeometry_p.cpp]:
```suggestion
  for ( int f : qgis::as_const( mFaces ) )
```

File: src/3d/terrain/qgstindemterraintilegeometry_p.h [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+    bool operator==( const QgsTriangularMeshTile &other ) const;
+
+    QgsRectangle tileExtent() const {return mExtent;}
+    QgsRectangle realTileExtent() const {return mRealExtent;}
+
+  private:
+    void init();
+
+    QgsTriangularMesh *mTriangularMesh;
```
Comment (nyalldawson) [inline src/3d/terrain/qgstindemterraintilegeometry_p.h]:
```suggestion
    QgsTriangularMesh *mTriangularMesh = nullptr;
```

File: src/3d/terrain/qgstindemterraintilegeometry_p.h [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    QVector<int> mFaces;
+
+    //! used to store the global index of the vertrices, the sum of the normales of associated triangles (weighted with area)
+    //! and the sum of the area of asociated triangles.
+    QVector<LocalVertex> mVertices;
+
+    //! used to retrive the local index from the glogal index
+    QHash<int, int> mLocalIndexFromGlobalIndex;
+
+    float mZmax;
```
Comment (nyalldawson) [inline src/3d/terrain/qgstindemterraintilegeometry_p.h]: Initialize

File: src/3d/terrain/qgstindemterraintilegeometry_p.h [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    float mZmin;
+};
+
+
+class QgsTinDemTerrainTileGeometry_p: public  Qt3DRender::QGeometry
+{
+  public:
+    explicit QgsTinDemTerrainTileGeometry_p( QgsTriangularMeshTile meshTile,
+        float verticaleScale, QNode *parent );
+    ~QgsTinDemTerrainTileGeometry_p()
```
Comment (nyalldawson) [inline src/3d/terrain/qgstindemterraintilegeometry_p.h]: remove empty destructor

File: src/app/3d/qgs3dmapconfigwidget.cpp [hunk 12]
```diff
@@ -139,6 +152,28 @@ void Qgs3DMapConfigWidget::apply()
       needsUpdateOrigin = true;
     }
   }
+  else if ( terrainType == QgsTerrainGenerator::Tin )
```
Comment (nyalldawson) [inline src/app/3d/qgs3dmapconfigwidget.cpp]: this should really be a `switch`

Comment (vcloarec) [reply src/app/3d/qgs3dmapconfigwidget.cpp]: jump bypasses variable initialization ?

File: src/3d/terrain/qgsterraintileloader_p.h [hunk 13]
```diff
@@ -30,6 +30,7 @@
 #include "qgschunkloader_p.h"
 
 #include <QImage>
+#include <QDebug>
```
Comment (m-kuhn) [inline src/3d/terrain/qgsterraintileloader_p.h]: Leftover?

File: src/app/3d/qgs3dmapconfigwidget.cpp (hunk 12 above)
Comment (m-kuhn) [reply src/app/3d/qgs3dmapconfigwidget.cpp]: add some `{}`

**Review by DelazJ — COMMENTED** [review]
Find some typos in the doc. 
Can you also add the [feature] tag for the changelog and user manual? Thanks

**Comment by nyalldawson** [comment]
The code looks good to me, but we need a review from @wonder-sk regarding the actual 3d implementation (that's out of my depth!)

**Comment by vcloarec** [comment]
After reflexion, I am not sure the used approach is the best. This approach consumes a lot of memory, I have to think for a better one when I will have time.

**Comment by nyalldawson** [comment]
Let's wait for @wonder-sk's thoughts... He may have some suggestions

**Comment by nirvn** [comment]
Snif.

**Comment by vcloarec** [comment]
> Snif.

Don't worry, it will come back! With a better form.

**Pull request closed without merging**